### PR TITLE
Update `flask-framework` extension

### PR DESCRIPTION
**Pull request by weiiwang01**
Some refactoring was performed, and some improvements were made over the `flask-framework` extension.

Changes include:

* Update the extension part name prefix from `flask/` to `flask-framework/`.
* Install `statsd-exporter` along with the Gunicorn server.
* Install ca-certificates in the ROCK image by default.
* Disable the WSGI path check when `services.flask.command` is overridden.
* Improve the quality of generated parts.

- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?

-----

### CODE REVIEW

**Comment by weiiwang01** [comment]
@sergiusens @tigarmo here's the pull request to bring only the latest `flask-framework` to the `main` branch.

**Review by tigarmo — APPROVED** [review]
Thanks a lot! To whoever is also reviewing this, this is the merge of the `feature/12f` branch "officially" into main (with some minor updates)

File: docs/reference/extensions.rst [hunk 1]
```diff
@@ -0,0 +1,19 @@
+Extensions
+**********
+
+Just as the Snapcraft extensions are designed to simplify Snap creation,
```
Comment (cmatsuoka) [inline docs/reference/extensions.rst:4]: Snap is capitalized here, should rock in line 23 above also be capitalized?

File: docs/reference/extensions.rst [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+Rockcraft extensions are crafted to expand and modify the user-provided
+rockcraft project file, aiming to minimise the boilerplate code when
+initiating a new rock.
```
Comment (cmatsuoka) [inline docs/reference/extensions.rst:7]: Or here.

File: rockcraft/commands/init.py [hunk 3]
```diff
@@ -52,6 +55,72 @@ def init(rockcraft_yaml_content: str) -> None:
 class InitCommand(AppCommand):
     """Initialize a rockcraft project."""
 
+    _INIT_TEMPLATES = {
+        "simple": textwrap.dedent(
+            """\
+                name: {name}
+                base: ubuntu@22.04 # the base environment for this ROCK
```
[the remark this answers is not in the document]
Comment (tigarmo) [reply rockcraft/commands/init.py:62]: @weiiwang01 can you fix this one please? The other ones should be fine ("rock" is correct)

[the remark this answers is not in the document]
Comment (tigarmo) [reply rockcraft/commands/init.py]: actually I just fixed this one myself

**Review by cmatsuoka — APPROVED** [review]

File: rockcraft/commands/init.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (weiiwang01) [reply rockcraft/commands/init.py]: @tigarmo thanks for fixing this problem! I will keep this in mind :)

File: docs/how-to/use-flask-extension.rst [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+Example:
+
+.. literalinclude:: code/use-flask-extension/example/rockcraft.yaml
+    :language: yaml
+
+Managing project files with the flask extension
+-----------------------------------------------
+
+By default the flask extension only includes the ``app.py``, ``static/``,
```
Comment (lengau) [inline docs/how-to/use-flask-extension.rst:19]:
```suggestion
By default the flask extension only includes ``app.py``, ``static/``,
```

File: docs/how-to/use-flask-extension.rst [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+``app/``, and ``templates/`` in the flask project. But you can overwrite this
```
Comment (lengau) [inline docs/how-to/use-flask-extension.rst:20]:
```suggestion
``app/``, and ``templates/`` in the flask project, but you can overwrite this
```

File: rockcraft/application.py [hunk 6]
```diff
@@ -37,7 +40,7 @@ class Rockcraft(Application):
 
     @override
     def _extra_yaml_transform(self, yaml_data: dict[str, Any]) -> dict[str, Any]:
-        return models.transform_yaml(self._work_dir, yaml_data)
+        return models.transform_yaml(Path.cwd(), yaml_data)
```
Comment (lengau) [inline rockcraft/application.py:43]: I'm not sure I understand this change. I don't oppose it, I'm just curious what it fixes.

File: rockcraft/commands/init.py [hunk 7]
```diff
@@ -52,6 +55,72 @@ def init(rockcraft_yaml_content: str) -> None:
 class InitCommand(AppCommand):
     """Initialize a rockcraft project."""
 
+    _INIT_TEMPLATES = {
```
Comment (lengau) [inline rockcraft/commands/init.py:58]: It's not needed here yet, but when this starts getting more complex maybe we can port Charmcraft's jinja-based implementation to a `TemplateService` in craft-application.

**Comment by sergiusens** [comment]
we have linter errors, are those expected?

File: rockcraft/application.py (hunk 6 above)
Comment (tigarmo) [reply rockcraft/application.py:43]: I should have documented this better, but the parameter that should be passed to `models.transform_yaml()` is the directory that contains the project file. In non-managed mode `self._work_dir` is the cwd but in managed mode [it isn't](https://github.com/canonical/craft-application/blob/main/craft_application/application.py#L143).
I think `self.project_dir` didn't exist yet when I made this change, we should use that instead.

**Comment by weiiwang01** [comment]
> we have linter errors, are those expected?

The issue was caused by an external website (https://yaml.org/) going down, which has been resolved now.

**Comment by tigarmo** [comment]
Will merge this as soon as I can merge [this hotfix](https://github.com/canonical/rockcraft/pull/511) (we prefer to merge hotfix branches as soon as possible)

**Pull request merged**